File: src/main/scala/algs/sort/MergeSort.scala
package algs.sort

import scala.annotation.tailrec

trait MergeSort {

  def sort(list: List[Int]): List[Int] = {
    val mid = list.length / 2
    mid match {
      case 0 => list
      case _ =>
        val (left, right) = list.splitAt(mid)
        merge(sort(left), sort(right))
    }
  }

  /*
    Note: `acc.reverse ::: left` and `lHead :: acc` is much
    cheaper than `acc ::: left` and `acc :+ lHead`
 */
  @tailrec
  private def merge(left: List[Int], right: List[Int], acc: List[Int] = Nil): List[Int] = {
    (left, right) match {
      case (_, Nil) => acc.reverse ::: left
      case (Nil, _) => acc.reverse ::: right
      case (lHead :: lTail, rHead :: rTail) =>
        if (lHead < rHead) merge(lTail, right, lHead :: acc)
        else merge(left, rTail, rHead :: acc)
    }
  }

}

File: algs/sort/merge_sort.py
def merge_sort(collection):
    split_merge(0, len(collection), collection)
    return collection


def split_merge(begin, end, collection):
    if (end - begin) > 0:
        middle = (end + begin) // 2
        split_merge(begin, middle, collection)
        split_merge(middle + 1, end, collection)
        merge(begin, middle, end, collection)


def merge(begin, middle, end, collection):
    left = collection[begin : middle + 1]
    right = collection[middle + 1 : end + 1]
    i = j = 0
    k = begin

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            collection[k] = left[i]
            i += 1
        else:
            collection[k] = right[j]
            j += 1
        k += 1

    if j == len(right):
        collection[k : end + 1] = left[i:]
